spec/activerecord-bitemporal/callbacks_spec.rb

```diff
@@ -111,61 +96,34 @@ class EmployeeWithBitemporalCallbacksLog < Employee
-          it { expect { subject }.not_to change(employee, :log) }
-          it { expect { subject }.not_to change { employee.job_title.log } }
-        end
-
-        context 'job_title with option `ignore_bitemporal_callbacks: true`' do
-          before { employee.job_title.bitemporal_option_merge!(ignore_bitemporal_callbacks: true) }
-          it { expect { subject }.to change(employee, :log).from([]).to(%i[before_bitemporal_update after_bitemporal_update]) }
-          it { expect { subject }.not_to change { employee.job_title.log } }
-        end
+      context 'employee and job_title have changes' do
+        let(:update_attributes) { { name: "Tom", job_title_attributes: { id: employee.job_title.id, name: "COO" } } }
+        it { expect { subject }.to change(employee, :log).from([]).to(%i[before_bitemporal_update after_bitemporal_update]) }
+        it { expect { subject }.to change { employee.job_title.log }.from([]).to(%i[before_bitemporal_update after_bitemporal_update]) }
       end
 
-      context 'with _assign_attribute' do
-        subject {
-          employee.name = "Tom"
-          employee.job_title.name = "COO"
-          employee.save!
-        }
+      context 'employee has changes and job_title does not have changes' do
+        let(:update_attributes) { { name: "Tom" } }
+        it { expect { subject }.to change(employee, :log).from([]).to(%i[before_bitemporal_update after_bitemporal_update]) }
+        it { expect { subject }.not_to change { employee.job_title.log } }
+      end
 
-        context 'without option `ignore_bitemporal_callbacks`' do
-          it { expect { subject }.to change(employee, :log).from([]).to(%i[before_bitemporal_update after_bitemporal_update]) }
-          it { expect { subject }.to change { employee.job_title.log }.from([]).to(%i[before_bitemporal_update after_bitemporal_update]) }
-        end
-
-        context 'employee with option `ignore_bitemporal_callbacks: true`' do
-          before { employee.bitemporal_option_merge!(ignore_bitemporal_callbacks: true) }
-          it { expect { subject }.not_to change(employee, :log) }
-          it { expect { subject }.to change { employee.job_title.log }.from([]).to(%i[before_bitemporal_update after_bitemporal_update]) }
-        end
-
-        context 'job_title with option `ignore_bitemporal_callbacks: true`' do
-          before { employee.job_title.bitemporal_option_merge!(ignore_bitemporal_callbacks: true) }
-          it { expect { subject }.to change(employee, :log).from([]).to(%i[before_bitemporal_update after_bitemporal_update]) }
-          it { expect { subject }.not_to change { employee.job_title.log } }
-        end
+      context 'employee does not have changes and job_title has changes' do
+        let(:update_attributes) { { job_title_attributes: { id: employee.job_title.id, name: "COO" } } }
+        it { expect { subject }.to change(employee, :log).from([]).to(%i[before_bitemporal_update after_bitemporal_update]) }
```

Comment:
I'm surprised that the callback is invoked for an unchanged record. Is this the natural behavior of Active Record?

Reply:
This behavior is the same behavior as ActiveRecord.
Even if ActiveRecord instance has no changes, update callbacks will be executed when save method is called.

Reply:
Oops, I misunderstood.